### PR TITLE
Various HTTP code cleanup.

### PR DESCRIPTION
**Pull request by millems**
1. Made SdkHttp*Request and SdkHttp*Response more closely follow the builder patterns we have used elsewhere.
2. Merged different SdkHttpUtils  and moved them to the utils package.
3. Use a common URI encoder across the SDK.
4. Added HttpStatusFamily for getting the family of a HTTP status code.

### CODE REVIEW

File: core/src/main/java/software/amazon/awssdk/auth/Aws4Signer.java [hunk 1]
```diff
@@ -272,24 +270,19 @@ public SdkHttpFullRequest presign(Context.BeforeTransmission execution,
      * .amazon.com/general/latest/gr/sigv4-create-canonical-request.html to
      * generate the canonical request.
      */
-    private String createCanonicalRequest(SdkHttpRequest request,
+    private String createCanonicalRequest(SdkHttpFullRequest.Builder request,
```
Comment (spfink) [inline core/src/main/java/software/amazon/awssdk/auth/Aws4Signer.java]: Don't know yet where this is being called but feels weird to pass in a mutable request object to signing rather than a fully built request.

File: core/src/test/java/software/amazon/awssdk/client/AsyncClientHandlerImplTest.java [hunk 2]
```diff
@@ -152,7 +153,8 @@ private void expectRetrievalFromMocks() {
     public AsyncClientConfiguration clientConfiguration() {
         MutableClientConfiguration mutableClientConfiguration = new MutableClientConfiguration()
                 .credentialsProvider(credentialsProvider)
-                .asyncHttpClient(httpClient);
+                .asyncHttpClient(httpClient)
+                .endpoint(URI.create("http://test.com"));
```
Comment (spfink) [inline core/src/test/java/software/amazon/awssdk/client/AsyncClientHandlerImplTest.java]: ?

File: core/src/test/java/software/amazon/awssdk/http/SdkHttpFullResponseAdapterTest.java [hunk 3]
```diff
@@ -23,28 +23,26 @@
...
+    private final SdkHttpFullRequest request = ValidSdkObjects.sdkHttpFullRequest().build();
 
     @Test
     public void adapt_SingleHeaderValue_AdaptedCorrectly() throws Exception {
-        SimpleSdkHttpFullResponse httpResponse = SimpleSdkHttpFullResponse.builder()
-                                                                          .header("FooHeader", "headerValue")
-                                                                          .build();
+        SdkHttpFullResponse httpResponse = SdkHttpFullResponse.builder()
+                                                              .header("FooHeader", "headerValue")
+                                                              .statusCode(200)
```
Comment (spfink) [inline core/src/test/java/software/amazon/awssdk/http/SdkHttpFullResponseAdapterTest.java]: Why is this required now?

File: http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java [hunk 4]
```diff
@@ -18,139 +18,211 @@
...
+        }
+
+        return standardizedPath.toString();
+    }
+
+    private Integer standardizePort(Integer port) {
+        Validate.isTrue(port == null || port >= -1,
+                        "Port must be positive (or null/-1 to indicate no port), but was '%s'", port);
+
+        if (port != null && port == -1) {
```
Comment (spfink) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java]: Do you need the null check?

File: services/s3/pom.xml [hunk 5]
```diff
@@ -93,5 +93,15 @@
             <artifactId>assertj-core</artifactId>
             <version>${assertj.version}</version>
         </dependency>
+        <dependency>
+            <groupId>log4j</groupId>
```
Comment (spfink) [inline services/s3/pom.xml]: Why do we need this dependency?

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java [hunk 6]
```diff
@@ -18,40 +18,100 @@
...
+     * URI string used for AWS request signing.
      *
-     * @return The service endpoint to which this request should be sent.
+     * <p>The URI's port will be missing (-1) when the {@link #port()} is the default port for the {@link #protocol()}. (80 for
+     * http and 443 for https). This is to reflect the fact that request signature does not include the port.</p>
+     *
+     * @return The URI for this request, formatted in the same way the AWS HTTP request signer uses the URI in the signature.
      */
-    URI getEndpoint();
+    default URI toUri() {
```
Comment (spfink) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java]: toUri doesn't seem like the right name here. getUri?

File: core/src/main/java/software/amazon/awssdk/auth/Aws4Signer.java (hunk 1 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/auth/Aws4Signer.java]: We pass in the immutable http request to the signer. The signer is implemented by calling toBuilder(), modifying it, then calling build(). This is one of the methods involved in the "modifying it" process.

Before this CR, the `SdkHttpRequest` was actually the builder as well. The builder implemented `SdkHttpRequest`. I changed that with this CR, so I changed the method signature

File: core/src/test/java/software/amazon/awssdk/client/AsyncClientHandlerImplTest.java (hunk 2 above)
Comment (millems) [reply core/src/test/java/software/amazon/awssdk/client/AsyncClientHandlerImplTest.java]: I made the endpoint components (protocol, host) required in the HTTP request so that the interface can guarantee they're always there. The endpoint in the client configuration is the one used to determine the protocol and host that goes into the HTTP request, so they're now required here, by proxy.

File: core/src/test/java/software/amazon/awssdk/http/SdkHttpFullResponseAdapterTest.java (hunk 3 above)
Comment (millems) [reply core/src/test/java/software/amazon/awssdk/http/SdkHttpFullResponseAdapterTest.java]: To avoid `SdkHttpResponse`'s `statusCode()` from returning `Optional<Integer>`, same as the endpoint above for request. The HTTP spec requires the service to always return the status code, so making it optional felt incorrect.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java (hunk 4 above)
Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java]: Yeah, if port is null, the unboxing involved in port == -1 would throw a NullPointerException.

File: services/s3/pom.xml (hunk 5 above)
Comment (millems) [reply services/s3/pom.xml]: I needed it to debug S3 tests that were failing, and I figured it would be a good thing to leave in. There was already a log4j.properties, so I figured logging was meant to be enabled. Let me know if you want me to remove it.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java (hunk 6 above)
Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java]: +1

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java [hunk 7]
```diff
@@ -40,11 +48,37 @@ public ApplyUserAgentStage(HttpClientDependencies dependencies) {
...
-        } else {
-            return request.header(HEADER_USER_AGENT, RuntimeHttpUtils.getUserAgent(clientConfig, null));
+        String userAgent = opts != null
+                           ? getUserAgent(clientConfig, opts.getClientMarker(RequestClientOptions.Marker.USER_AGENT))
+                           : getUserAgent(clientConfig, null);
+
+        return request.header(HEADER_USER_AGENT, userAgent);
+    }
+
+    private String getUserAgent(ClientConfiguration config, final String userAgentMarker) {
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: is userAgentMarker the requestLevel user agent? If so can we rename it?

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java [hunk 8]
```diff
@@ -40,11 +48,37 @@ public ApplyUserAgentStage(HttpClientDependencies dependencies) {
...
+        ClientOverrideConfiguration overrideConfig = config.overrideConfiguration();
+        String userDefinedPrefix = overrideConfig.advancedOption(AdvancedClientOption.USER_AGENT_PREFIX);
+        String userDefinedSuffix = overrideConfig.advancedOption(AdvancedClientOption.USER_AGENT_SUFFIX);
+        String awsExecutionEnvironment = AwsSystemSetting.AWS_EXECUTION_ENV.getStringValue().orElse(null);
+
+        StringBuilder userAgent = new StringBuilder(software.amazon.awssdk.utils.StringUtils.trimToEmpty(userDefinedPrefix));
+
+        if (!UserAgentUtils.getUserAgent().equals(userDefinedPrefix)) {
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: This check seems unnecessary.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java [hunk 9]
```diff
@@ -199,10 +199,10 @@ private void beforeExecute() {
 
         private CompletableFuture<Response<OutputT>> doExecute() throws Exception {
             if (isRetry()) {
-                resetRequestInputStream(request.getContent());
+                resetRequestInputStream(request.content().orElse(null));
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java]: Can you do an ifPresent so you don't have to do a null check in the method?

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java [hunk 10]
```diff
@@ -199,10 +199,10 @@ private void beforeExecute() {
...
             }
 
-            markInputStream(request.getContent());
+            markInputStream(request.content().orElse(null));
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java]: Same here, can you use ifPresent.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java [hunk 11]
```diff
@@ -22,30 +22,31 @@
...
 
     private boolean shouldPutParamsInBody(SdkHttpFullRequest.Builder input,
                                           RequestExecutionContext context) {
         return notSimpleDb(context) &&
-               input.getHttpMethod() == SdkHttpMethod.POST &&
-               input.getContent() == null &&
-               input.getParameters() != null &&
-               input.getParameters().size() > 0;
+               input.method() == SdkHttpMethod.POST &&
+               input.content() == null &&
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java]: Is content an optional now? Do we need to check isPresent?

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java [hunk 12]
```diff
@@ -22,30 +22,31 @@
...
+               input.rawQueryParameters() != null &&
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java]: I think there is a isNullOrEmpty method for maps somewhere.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/RetryableStage.java [hunk 13]
```diff
@@ -170,11 +170,11 @@ private void beforeExecute() throws InterruptedException {
 
         private Response<OutputT> doExecute() throws Exception {
             if (isRetry()) {
-                resetRequestInputStream(request.getContent());
+                resetRequestInputStream(request.content().orElse(null));
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/RetryableStage.java]: ifPresent

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/RetryableStage.java [hunk 14]
```diff
@@ -170,11 +170,11 @@ private void beforeExecute() throws InterruptedException {
...
                 pauseBeforeRetry();
             }
 
-            markInputStream(request.getContent());
+            markInputStream(request.content().orElse(null));
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/RetryableStage.java]: ifPresent

File: core/src/main/java/software/amazon/awssdk/runtime/auth/SignerProviderContext.java [hunk 15]
```diff
@@ -15,21 +15,18 @@
...
-import java.net.URI;
 import software.amazon.awssdk.RequestConfig;
 import software.amazon.awssdk.annotation.SdkProtectedApi;
 import software.amazon.awssdk.http.SdkHttpFullRequest;
 
 @SdkProtectedApi
 public class SignerProviderContext {
 
-    private final URI uri;
     private final boolean isRedirect;
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/runtime/auth/SignerProviderContext.java]: Do we need isRedirect either?

File: core/src/main/java/software/amazon/awssdk/auth/AbstractAwsSigner.java [hunk 16]
```diff
@@ -224,32 +220,24 @@ protected String getCanonicalizedQueryString(Map<String, List<String>> parameter
...
+            final String encodedParamName = SdkHttpUtils.urlEncode(entry.getKey());
             final List<String> paramValues = entry.getValue();
-            final List<String> encodedValues = new ArrayList<String>(
-                    paramValues.size());
+            final List<String> encodedValues = new ArrayList<>(paramValues.size());
             for (String value : paramValues) {
-                encodedValues.add(SdkHttpUtils.urlEncode(value, false));
+                String encodedValue = SdkHttpUtils.urlEncode(value);
+
+                // Null values should be treated as empty for the purposes of signing, not missing.
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/auth/AbstractAwsSigner.java]: This is the existing behavior correct? Looking at the Sigv4 spec this is correct but just making sure we do it right in V1 as well.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java [hunk 17]
```diff
@@ -18,139 +18,211 @@
...
-        private String resourcePath;
-        @ReviewBeforeRelease("Do we need linked hash map here?")
+        private String protocol;
+        private String host;
+        private Integer port;
+        private String path;
         private Map<String, List<String>> queryParameters = new LinkedHashMap<>();
-        private URI endpoint;
         private SdkHttpMethod httpMethod;
+        private Map<String, List<String>> headers = new HashMap<>();
```
Comment (shorea) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java]: Should we use TreeMap here to be consistent?

File: core/src/main/java/software/amazon/awssdk/auth/Aws4Signer.java [hunk 18]
```diff
@@ -272,24 +270,19 @@ public SdkHttpFullRequest presign(Context.BeforeTransmission execution,
...
-                .append(SignerConstants.LINE_SEPARATOR)
-                .append(getCanonicalizedHeaderString(request))
-                .append(SignerConstants.LINE_SEPARATOR)
-                .append(getSignedHeadersString(request)).append(SignerConstants.LINE_SEPARATOR)
-                .append(contentSha256)
-                .toString();
+        final String canonicalRequest = request.method().toString() +
+                                        SignerConstants.LINE_SEPARATOR +
+                                        // This would optionally double url-encode the resource path
+                                        getCanonicalizedResourcePath(request.encodedPath(), doubleUrlEncode) +
```
[the remark this answers is not in the document]
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/auth/Aws4Signer.java]: Sure.

File: core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsRetryPolicy.java [hunk 19]
```diff
@@ -33,7 +34,7 @@ public boolean shouldRetry(int retriesAttempted, CredentialsEndpointRetryParamet
         }
 
         Integer statusCode = retryParams.getStatusCode();
-        if (statusCode != null && statusCode >= 500 && statusCode < 600) {
+        if (statusCode != null && HttpStatusFamily.of(statusCode) == HttpStatusFamily.SERVER_ERROR) {
```
[the remark this answers is not in the document]
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsRetryPolicy.java]: Is the desire to not do the `==`?

File: core/src/main/java/software/amazon/awssdk/http/SdkHttpFullRequestAdapter.java [hunk 20]
```diff
@@ -35,18 +36,20 @@ public static SdkHttpFullRequest toHttpFullRequest(Request<?> request) {
 
     public static SdkHttpFullRequest.Builder toMutableHttpFullRequest(Request<?> request) {
         return SdkHttpFullRequest.builder()
-                                 .content(request.getContent())
-                                 .httpMethod(SdkHttpMethod.fromValue(request.getHttpMethod().name()))
+                                 .protocol(request.getEndpoint().getScheme())
```
[the remark this answers is not in the document]
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/SdkHttpFullRequestAdapter.java]: The reason I was a bit hesitant to do that is because a URI has more than just a scheme, host and port. We'd probably want to decode the path, query parameters, etc. That felt like overkill in our case.

File: core/src/main/java/software/amazon/awssdk/http/async/SyncResponseHandlerAdapter.java [hunk 21]
```diff
@@ -79,7 +80,10 @@ public T complete() {
         try {
             // Once we've buffered all the content we can invoke the response handler
             if (baos != null) {
-                httpResponse.content(new ByteArrayInputStream(baos.toByteArray()));
+                // Ignore aborts - we already have all of the content.
+                ByteArrayInputStream content = new ByteArrayInputStream(baos.toByteArray());
+                AbortableInputStream abortableContent = new AbortableInputStream(content, () -> { });
```
[the remark this answers is not in the document]
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/async/SyncResponseHandlerAdapter.java]: okay

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java (hunk 7 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: Not sure, I copied this code from the `SdkRuntimeUtils`. I'll look into it.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java (hunk 8 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: I copied this. I'll look into it.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java (hunk 9 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java]: +1

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java (hunk 10 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/AsyncRetryableStage.java]: +1

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java (hunk 11 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java]: It's not optional in the builder. It's whatever they set in the set method. That's to avoid everything in the "get" from builders having to be optional. LMK if you think it should be the other way.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java (hunk 12 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java]: I'll see if I can find it!

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java (hunk 11 above)
Comment (shorea) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java]: +1, didn't realize this was on the builder.

File: core/src/main/java/software/amazon/awssdk/auth/AbstractAwsSigner.java (hunk 16 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/auth/AbstractAwsSigner.java]: Yeah, we do it right in V1. We had a few ways of doing encoding in V1, and the one this one used encoded it in that way. I made all of the encoding use the same path with this, so I had to make the tweak here.

File: core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsRetryPolicy.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (shorea) [reply core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsRetryPolicy.java]: Mainly just shorter.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java (hunk 17 above)
Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java]: I'd actually vote for `LinkedHashMap` so that when they do a set and then a get, the order (if relevant) is preserved. The contract on the "get" is the same value as being set. It'll be re-shuffled when the message itself is built to match the request's contract.

File: core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsRetryPolicy.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsRetryPolicy.java]: Okay, how about `HttpStatusFamily.of(statusCode).isServerError()`? Should we have similar methods for all of the other families?

File: core/src/main/java/software/amazon/awssdk/http/async/SyncResponseHandlerAdapter.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/async/SyncResponseHandlerAdapter.java]: It's a bit too long, so I have to newline the parameter. Do you mind if I leave it the same?

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java (hunk 8 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: Why does it seem uneccessary? I don't understand the format of the user agent well enough to know why it's not needed.

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java (hunk 12 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MoveParametersToBodyStage.java]: Couldn't find one for anything other than lists. Added one to `CollectionUtils`.

File: core/src/main/java/software/amazon/awssdk/runtime/auth/SignerProviderContext.java (hunk 15 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/runtime/auth/SignerProviderContext.java]: We don't need the context at all anymore. Think we should save that cleanup for the signer refactor?

File: core/src/main/java/software/amazon/awssdk/http/async/SyncResponseHandlerAdapter.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (shorea) [reply core/src/main/java/software/amazon/awssdk/http/async/SyncResponseHandlerAdapter.java]: Yeah that's fine

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java (hunk 8 above)
Comment (shorea) [reply core/src/main/java/software/amazon/awssdk/http/pipeline/stages/ApplyUserAgentStage.java]: So if I understand correctly we are just checking that the user supplied prefix doesn't exactly match our internal user agent. That seems like an error on the customers part if they do that for whatever reason.

File: core/src/main/java/software/amazon/awssdk/runtime/auth/SignerProviderContext.java (hunk 15 above)
Comment (shorea) [reply core/src/main/java/software/amazon/awssdk/runtime/auth/SignerProviderContext.java]: +1, sounds good

File: http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java (hunk 17 above)
Comment (shorea) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java]: Didn't grok that, could you give an example?

File: services/s3/pom.xml (hunk 5 above)
Comment (shorea) [reply services/s3/pom.xml]: Don't you only need the slf4j one?

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/SqsExecutionInterceptor.java [hunk 22]
```diff
@@ -42,16 +41,11 @@
...
 
         // If the request is using a non-standard endpoint, then
         // alter it to use the corresponding, standard endpoint
-        if (NONSTANDARD_ENDPOINT_MAP.containsKey(endpoint.getHost())) {
-            String newHost = NONSTANDARD_ENDPOINT_MAP.get(endpoint.getHost());
-            String newEndpoint = endpoint.toString().replaceFirst(endpoint.getHost(), newHost);
-            return request.toBuilder()
-                          .endpoint(URI.create(newEndpoint))
-                          .build();
+        if (NONSTANDARD_ENDPOINT_MAP.containsKey(request.host())) {
```
Comment (shorea) [inline services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/SqsExecutionInterceptor.java]: I'd be in favor of removing this interceptor altogether.

File: utils/src/main/java/software/amazon/awssdk/utils/builder/SdkBuilder.java [hunk 23]
```diff
@@ -55,9 +54,8 @@
      * @return B the mutated builder instance
      */
     @SuppressWarnings("unchecked")
-    @ReviewBeforeRelease("Why does the user have to return a builder if they're just modifying it? "
-                         + "That makes this harder to use.")
-    default B apply(UnaryOperator<B> function) {
-        return function.apply((B) this);
+    default B apply(Consumer<B> function) {
```
Comment (shorea) [inline utils/src/main/java/software/amazon/awssdk/utils/builder/SdkBuilder.java]: Should we rename this to accept to match the Consumer interface? Also no longer a function

File: services/s3/pom.xml (hunk 5 above)
Comment (millems) [reply services/s3/pom.xml]: Let me triple-check, I thought the slf4j-log4j12 was just the bridge.

File: http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java (hunk 17 above)
Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java]: Sure! The javadoc on the builder says something like builder.headers() is "the same value as the one set on builder.headers(Collection<String>) (and the other methods)".

LinkedHashMap makes it more true by ensuring the elements remain in order:
```Java
HashMap<String> headers = ...;
HashMap<String> headersFromBuilder = builder.headers(headers).headers();

assertThat(headersFromBuilder).isEqualTo(headers); 
// True with LinkedHashMap, maybe False with standard hash map / sorted map.
```

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullRequest.java [hunk 24]
```diff
@@ -21,17 +21,20 @@
...
 
 /**
- * Represents a request being sent to an Amazon Web Service, including the
- * parameters being sent as part of the request, the endpoint to which the
- * request should be sent, etc.
+ * An immutable HTTP request with a possible HTTP body.
  *
- * <p>This class should not be implemented outside of the SDK.</p>
+ * <p>All implementations of this interface MUST be immutable. Instead of implementing this interface, consider using
+ * {@link #builder()} to create an instance.</p>
```
[the remark this answers is not in the document]
Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullRequest.java]: :+1:

File: http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullResponse.java [hunk 25]
```diff
@@ -15,32 +15,27 @@
...
+import software.amazon.awssdk.annotation.SdkPublicApi;
 import software.amazon.awssdk.utils.builder.CopyableBuilder;
 import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
 
 /**
- * Generic representation of an HTTP response.
+ * An immutable HTTP response with a possible HTTP body.
  *
- * TODO should we allow HTTP impl authors to implement this or should they only use the builder?
+ * <p>All implementations of this interface MUST be immutable. Instead of implementing this interface, consider using
```
[the remark this answers is not in the document]
Comment (millems) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullResponse.java]: :+1:

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/RepeatableInputStreamRequestEntity.java [hunk 26]
```diff
@@ -78,17 +79,17 @@ public RepeatableInputStreamRequestEntity(final SdkHttpFullRequest request) {
          * buffer the entire stream contents into memory to determine
          * the content length.
          */
-        long contentLength = request.getFirstHeaderValue("Content-Length")
-                .map(this::parseContentLength)
-                .orElse(-1L);
+        long contentLength = SdkHttpUtils.firstMatchingHeader(request.headers(), "Content-Length")
```
[the remark this answers is not in the document]
Comment (millems) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/RepeatableInputStreamRequestEntity.java]: I moved it to a utility method because it felt useful even if we don't have access to an HTTP request/response. For example, code that is using the builder or HTTP plugin code before it has created the request/response can use the functionality now.

The reason I didn't add a convenience method onto both the builder and request/response objects to delegate to the util class:
- There were already a lot of other methods on the builder/request already.
- There are a few things in SdkHttpUtils that could make sense on the interfaces, and I was worried about the subjectivity of "should we add a convenience method for this to the builder/request interfaces?" Granted, I did add `getUri` to the request, so I'm being a bit hypocritical here. That was to avoid the utils -> http-spi dependency.

Let me know if you feel like it's worth keeping.

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java [hunk 27]
```diff
@@ -128,24 +93,24 @@ private RequestCallable(HttpURLConnection connection, SdkHttpFullRequest request
...
             int responseCode = connection.getResponseCode();
-            InputStream content = responseCode < 400 ? connection.getInputStream() : connection.getErrorStream();
+            boolean isErrorResponse = HttpStatusFamily.of(responseCode).isOneOf(CLIENT_ERROR, SERVER_ERROR);
+            InputStream content = !isErrorResponse ? connection.getInputStream() : connection.getErrorStream();
+
             return SdkHttpFullResponse.builder()
                                       .statusCode(responseCode)
                                       .statusText(connection.getResponseMessage())
-                                      .content(content)
+                                      .content(new AbortableInputStream(content, () -> { /* TODO: Don't ignore abort? */ }))
```
[the remark this answers is not in the document]
Comment (millems) [reply http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java]: I wasn't entirely confident that `connection.disconnect()` was the correct thing to do in this case without making sure I understood exactly how the client and URL connection worked. Keeping the behavior and adding a TODO felt safer at the time. I'll look into it now.

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/SqsExecutionInterceptor.java (hunk 22 above)
Comment (millems) [reply services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/SqsExecutionInterceptor.java]: Sure, if it's safe. Not sure what it's for.

File: utils/src/main/java/software/amazon/awssdk/utils/builder/SdkBuilder.java (hunk 23 above)
Comment (millems) [reply utils/src/main/java/software/amazon/awssdk/utils/builder/SdkBuilder.java]: `builder.accept(b -> b.endpoint("foo"))` just doesn't look quite right to me. `apply` feels okay to me, since we're applying a mutation to the builder. I can't quite think of something better.

Renamed the parameter.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/RepeatableInputStreamRequestEntity.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (shorea) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/RepeatableInputStreamRequestEntity.java]: I dunno, I feel like the interfaces should be rich enough and useful enough for us and SPI implementors work with. +1 for the utility but I think it would also be nice on the interface.
